dashboard/serializers.py

```diff
@@ -38,6 +39,29 @@ def serialize_enrollments(cls, mmtrack, enrollments):
         return serialized_enrollments
 
     @classmethod
+    def count_courses_passed(cls, program, mmtrack):
+        """
+        Returns number of passed courses.
+
+        Args:
+            program (Program): A program object.
+            mmtrack (dashboard.utils.MMTrack): a instance of all user information about a program
+        Returns:
+            num_complete_courses (int): A number of passed courses.
+        """
+        num_courses_passed = 0
+        for course in program.course_set.all():
+            course_info_dict = get_info_for_course(course, mmtrack)
+            if course_info_dict:
+                expected_statuses = list(
+                    filter(lambda c: c['status'] in [CourseStatus.PASSED], course_info_dict["runs"])
+                )
+                if len(expected_statuses) > 0:
+                    num_courses_passed += 1
+
+        return num_courses_passed
```

Comment:
it doesn't count course runs twice. below i pasted some commands and their results. you can try them yourself if you like. `self.edx_key_course_map` is a dict that maps a course run edx_course_key to the parent course's id. `count_courses_passed` creates a set of the `Course.id`'s that have been passed, ensuring uniqueness. again, i want to stress that we shouldn't confuse "changing existing code to meet new requirements" with "refactoring". those are not the same thing. if you are concerned with some particular risks with the change i'm suggesting to `MMTrack`, please let me know what that is specifically. i can identify 2 specific problems with the current approach. (1) we are using `get_info_for_course` for a purpose that it wasn't intended, and that function is performing a lot of work that doesn't need to be done. (2) it should not be the responsibility of a serializer class to calculate how many courses a student has passed. `MMTrack` is the class that determines if a single course run passed, so it makes sense that it should be able to tell us how many courses have been passed. if you still have questions/concerns, it might be better to chat in person

```shell
(micromasters) ~/d/micromasters ❯❯❯ git status --short                                                       
 M dashboard/serializers.py
 M dashboard/utils.py
(micromasters) ~/d/micromasters ❯❯❯ curl -s -X POST "http://192.168.99.100:9100/micromasters/_search?q=profile.username:fake.kenzi.cunningham" | jq '.["hits"]["hits"][0]["_source"]["program"]["num_courses_passed"]'
2
(micromasters) ~/d/micromasters ❯❯❯ mmmanage alter_data --action=add_past_passed_run --username=staff --program-title='Analog' --course-title='100' --add-if-exists
Course run changed: (id: 2479, edx_course_key: course-v1:MITx+Analog+Learning+100+Jan_2015)
User: staff; Course: Analog Learning 100
(micromasters) ~/d/micromasters ❯❯❯ docker-compose run web ./manage.py recreate_index                        
(micromasters) ~/d/micromasters ❯❯❯ curl -s -X POST "http://192.168.99.100:9100/micromasters/_search?q=profile.username:fake.kenzi.cunningham" | jq '.["hits"]["hits"][0]["_source"]["program"]["num_courses_passed"]'
2
(micromasters) ~/d/micromasters ❯❯❯ mmmanage alter_data --action=set_course_to_passed --username=kenzi --program-title='Analog' --course-title='300'
Course run changed: (id: 2483, edx_course_key: course-v1:MITx+Analog+Learning+300+Oct_2016)
User: fake.kenzi.cunningham; Course: Analog Learning 300
(micromasters) ~/d/micromasters ❯❯❯ docker-compose run web ./manage.py recreate_index                        
(micromasters) ~/d/micromasters ❯❯❯ curl -s -X POST "http://192.168.99.100:9100/micromasters/_search?q=profile.username:fake.kenzi.cunningham" | jq '.["hits"]["hits"][0]["_source"]["program"]["num_courses_passed"]'
3
(micromasters) ~/d/micromasters ❯❯❯ mmmanage alter_data --action=add_past_passed_run --username=staff --program-title='Analog' --course-title='300' --add-if-exists
Course run changed: (id: 2485, edx_course_key: course-v1:MITx+Analog+Learning+300+Jan_2015)
User: staff; Course: Analog Learning 300
(micromasters) ~/d/micromasters ❯❯❯ docker-compose run web ./manage.py recreate_index                        
(micromasters) ~/d/micromasters ❯❯❯ curl -s -X POST "http://192.168.99.100:9100/micromasters/_search?q=profile.username:fake.kenzi.cunningham" | jq '.["hits"]["hits"][0]["_source"]["program"]["num_courses_passed"]'
3
```